storage/shard.go

```diff
@@ -227,46 +236,103 @@ func (s *dbShard) ReadEncoded(
 }
 
+// getEntryWithLock returns the entry for a given id while holding a read lock or a write lock.
+func (s *dbShard) getEntryWithLock(id string) (*dbShardEntry, *list.Element, bool) {
+	elem, exists := s.lookup[id]
+	if !exists {
+		return nil, nil, false
+	}
+	return elem.Value.(*dbShardEntry), elem, true
+}
+
 func (s *dbShard) writableSeries(id string) (databaseSeries, uint64, writeCompletionFn) {
 	s.RLock()
-	if entry, exists := s.lookup[id]; exists {
+	if entry, _, exists := s.getEntryWithLock(id); exists {
 		entry.incrementWriterCount()
 		s.RUnlock()
-		return entry.series, entry.uniqueIndex, entry.decrementWriterCount
+		return entry.series, entry.index, entry.decrementWriterCount
 	}
 	s.RUnlock()
 
 	s.Lock()
-	if entry, exists := s.lookup[id]; exists {
+	if entry, _, exists := s.getEntryWithLock(id); exists {
 		entry.incrementWriterCount()
 		s.Unlock()
 		// During Rlock -> Wlock promotion the entry was inserted
-		return entry.series, entry.uniqueIndex, entry.decrementWriterCount
+		return entry.series, entry.index, entry.decrementWriterCount
 	}
 	bs := bootstrapNotStarted
 	if s.newSeriesBootstrapped {
 		bs = bootstrapped
 	}
-	series := newDatabaseSeries(id, bs, s.opts)
-	elem := s.list.PushBack(series)
 	entry := &dbShardEntry{
-		series:      series,
-		uniqueIndex: s.uniqueIndex.nextUniqueIndex(),
-		elem:        elem,
-		curWriters:  1,
+		series:     newDatabaseSeries(id, bs, s.opts),
+		index:      s.increasingIndex.nextIndex(),
+		curWriters: 1,
 	}
-	s.lookup[id] = entry
+	elem := s.list.PushBack(entry)
+	s.lookup[id] = elem
 	s.Unlock()
 
-	return entry.series, entry.uniqueIndex, entry.decrementWriterCount
+	return entry.series, entry.index, entry.decrementWriterCount
+}
+
+func (s *dbShard) FetchBlocks(
+	ctx context.Context,
+	id string,
+	starts []time.Time,
+) []FetchBlockResult {
+	s.RLock()
+	entry, _, exists := s.getEntryWithLock(id)
+	s.RUnlock()
+	if !exists {
+		return nil
+	}
+	return entry.series.FetchBlocks(ctx, starts)
+}
+
+func (s *dbShard) FetchBlocksMetadata(
+	ctx context.Context,
+	limit int64,
+	pageToken int64,
+	includeSizes bool,
+) ([]FetchBlocksMetadataResult, *int64) {
+	var (
+		res            = make([]FetchBlocksMetadataResult, 0, limit)
+		pNextPageToken *int64
+	)
+	s.forEachShardEntry(true, func(entry *dbShardEntry) error {
+		if int64(entry.index) < pageToken {
+			return nil
+		}
+
+		// Create a temporary context here so the stream readers can be returned to
+		// pool after we finish fetching the metadata for this series.
+		tmpCtx := s.opts.GetContextPool().Get()
+		blocksMetadata := entry.series.FetchBlocksMetadata(tmpCtx, includeSizes)
+		tmpCtx.Close()
+		res = append(res, blocksMetadata)
+
+		return nil
+	}, func(entry *dbShardEntry) bool {
+		// Break out of the iteration loop once we've accumulated enough entries.
+		if int64(len(res)) < limit {
+			return false
+		}
+		nextPageToken := int64(entry.index)
+		pNextPageToken = &nextPageToken
+		return true
```

Comment:
Hey so what if the stop function doesn't get called because we run out of results, will `nextPageToken` get correctly set?  If so how?  We need to return a next token that's going to not cause another read, i.e. a non-existent index I would think.